### ISSUE TITLE
os/exec: The TestImplicitPWD test fails when running under a soft link folder.

### ISSUE THREAD

**Issue by zhangfannie** [golang/go#52537]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version devel go1.19-96c8cc7fea Sun Apr 24 01:22:21 2022 +0000 linux/arm64
</pre>

### Does this issue reproduce with the latest release?
yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE="auto"
GOARCH="arm64"
GOBIN=""
GOCACHE="/home/fanzha02/.cache/go-build"
GOENV="/home/fanzha02/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="arm64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/fanzha02/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/fanzha02/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/home/fanzha02/sharefolder/golang"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/home/fanzha02/sharefolder/golang/pkg/tool/linux_arm64"
GOVCS=""
GOVERSION="devel go1.19-96c8cc7fea Sun Apr 24 01:22:21 2022 +0000"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/fanzha02/sharefolder/golang/src/go.mod"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build1974904193=/tmp/go-build -gno-record-gcc-switches"

</pre></details>

### What did you do?

Run the TestImplicitPWD test under a soft link folder.

### What did you expect to see?

pass

### What did you see instead?
~/sharefolder/golang/src/os/exec$ ../../../bin/go test -run=TestImplicitPWD
--- FAIL: TestImplicitPWD (0.00s)
    --- FAIL: TestImplicitPWD/empty (0.00s)
        exec_posix_test.go:161: in

            `pwd` reported
                /mnt/share/homes/fanzha02/golang/src/os/exec
        exec_posix_test.go:163: want
                /home/fanzha02/sharefolder/golang/src/os/exec
    --- FAIL: TestImplicitPWD/PWD (0.00s)
        exec_posix_test.go:161: in
                /home/fanzha02/sharefolder/golang/src/os/exec
            `pwd` reported
                /mnt/share/homes/fanzha02/golang/src/os/exec
        exec_posix_test.go:163: want
                /home/fanzha02/sharefolder/golang/src/os/exec
    --- FAIL: TestImplicitPWD/dot (0.02s)
        exec_posix_test.go:161: in
                .
            `pwd` reported
                /mnt/share/homes/fanzha02/golang/src/os/exec
        exec_posix_test.go:163: want
                /home/fanzha02/sharefolder/golang/src/os/exec
    --- FAIL: TestImplicitPWD/dotdot (0.02s)
        exec_posix_test.go:161: in
                ..
            `pwd` reported
                /mnt/share/homes/fanzha02/golang/src/os
        exec_posix_test.go:163: want
                /home/fanzha02/sharefolder/golang/src/os
    --- FAIL: TestImplicitPWD/PWDdotdot (0.02s)
        exec_posix_test.go:161: in
                /home/fanzha02/sharefolder/golang/src/os/exec/..
            `pwd` reported
                /mnt/share/homes/fanzha02/golang/src/os
        exec_posix_test.go:163: want
                /home/fanzha02/sharefolder/golang/src/os
FAIL
exit status 1

**Comment by zhangfannie** [comment]
The folder where I run the test is a soft link folder.
```
 > ls -all ~/sharefolder
 /home/fanzha02/sharefolder -> /mnt/share/homes/fanzha02
```
The `TestImplicitPWD` test is to compare whether the output of `os.Getwd()` and the output of `exec.Command("pwd")` are equal.  When run it **under a soft link folder**, the `os.Getwd()` returns the directory of soft link folder, it is `/home/fanzha02/sharefolder`, the `exec.Command("pwd")` returns the directory of source folder, it is `/mnt/share/homes/fanzha02`, different outputs cause the test to fail.

I am not familar with the destination of the `TestImplicitPWD` test, **is this failure iexpected?** Thank you.

**Comment by zhangfannie** [comment]
@bcmills

**Comment by bcmills** [comment]
Heh, looks like `/bin/pwd` doesn't actually behave the same as the `pwd` built-in:
```
/tmp$ ln -s ~/go-review goroot-link

/tmp$ cd goroot-link/src/os/exec

/tmp/goroot-link/src/os/exec$ pwd
/tmp/goroot-link/src/os/exec

/tmp/goroot-link/src/os/exec$ /bin/pwd
/usr/local/google/home/bcmills/go-review/src/os/exec
```

This is fixed by #401934, but a more local fix seems possible.
